smartconfig/entry.py

```diff
@@ -0,0 +1,139 @@
+            raise ConfigurationKeyError(f"Entry {path} doesn't define any {attribute} entry.")
+
+        return registry.global_configuration[path][attribute]
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any], path: Optional[str] = None) -> type:
+        """
+        Add special attribute to the new entry.
+
+        Args:
+            path: Custom path used for this entry instead of the module name.
+        """
+        names = chain(dict_.keys(), dict_.get('__annotations__', {}).keys())
+        dict_[f"{cls.__name__}__defined_entries"] = {name for name in names if not name.startswith('_')}
+
+        dict_[f"{cls.__name__}__path_override"] = path
+
+        return super().__new__(cls, name, bases, dict_)
+
+    def _register_entry(cls) -> None:
+        """Set the `__path` attribute and register the entry."""
+        cls.__path = cls.__path_override or cls.__module__
+        if cls.__path in registry.configuration_for_module:
+            raise DuplicateConfiguration(f"An entry at {cls.__path} already exists.")  # TODO: Add an FAQ link.
+
+        configuration = {
+            key: value for key, value in cls.__dict__.items() if not key.startswith('_')
+        }
+
+        if cls.__path not in registry.global_configuration:
+            registry.global_configuration[cls.__path] = configuration
+        # We already have some overrides for this path.
+        else:
+            for key, value in configuration.items():
+                # We only write values that aren't already defined.
+                if key not in registry.global_configuration[cls.__path]:
+                    registry.global_configuration[cls.__path][key] = value
+
+        registry.configuration_for_module[cls.__path] = cls
+
+    def _check_undefined_entries(cls) -> None:
+        """Raise `ConfigurationKeyError` if any attribute doesn't have a concrete value."""
+        for attribute in cls.__defined_entries:
+            if attribute not in registry.global_configuration[cls.__path]:
+                raise ConfigurationKeyError(f"Entry {attribute} isn't defined.")
+
+    def _get_attribute_path(cls, attribute_name: str) -> Tuple[str, str]:
+        """
+        Returns which path to use to lookup the attribute.
+
+        By default `(cls.__path, attribute_name)` will be returned.
+
+        Args:
+            attribute_name: The name of the attribute to lookup.
+
+        Returns:
+            A tuple with the configuration path as the first element and the element name to lookup.
+        """
+        return cls.__path, attribute_name
+
+    def __init__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any], path: Optional[str] = None):
+        """
+        Initialize the new entry.
+
+        Raises:
+            DuplicateConfiguration: An entry is already registered for this path, use `__path_override`.
+            ConfigurationKeyError: An attribute doesn't have a concrete value.
```

Comment:
This will be raised if an attribute is defined through a typehint with any (concrete?) value and the config doesn't provide a value either. I'm no good English speaker, what word could fit nicely here?

Reply:
Using terminology popular in C, the attribute is declared but not defined. You could change it to "doesn't have a defined value." PEP 526 calls it a "variable without initial value" and "variable without default."